storage/parquet/datasource.go

```diff
@@ -0,0 +1,313 @@
+				case parquetformat.Type_INT96:
+					elements = append(elements, it.int96s[it.dataIterator])
+				case parquetformat.Type_FLOAT:
+					elements = append(elements, it.float32s[it.dataIterator])
+				case parquetformat.Type_DOUBLE:
+					elements = append(elements, it.float64s[it.dataIterator])
+				case parquetformat.Type_BYTE_ARRAY, parquetformat.Type_FIXED_LEN_BYTE_ARRAY:
+					elements = append(elements, it.byteArrays[it.dataIterator])
+				default:
+					panic("unknown type")
+				}
+			}
+
+			it.dataIterator++
+		} else {
+			elements = append(elements, nil)
+		}
+
+		it.descriptionIterator++
+
+		if it.col.MaxR() == 0 {
+			break;
+		}
+	}
+
+	if it.col.MaxR() > 0 {
+		s = elements
+	}
+	return s, nil
+}
+
+func NewDataSourceBuilderFactory() physical.DataSourceBuilderFactory {
+	return physical.NewDataSourceBuilderFactory(
+		func(ctx context.Context, matCtx *physical.MaterializationContext, dbConfig map[string]interface{}, filter physical.Formula, alias string) (execution.Node, error) {
+			path, err := config.GetString(dbConfig, "path")
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get path")
+			}
+
+			return &DataSource{
+				path:  path,
+				alias: alias,
+			}, nil
+		},
+		nil,
+		availableFilters,
+		metadata.BoundedFitsInLocalStorage,
+	)
+}
+
+// NewDataSourceBuilderFactoryFromConfig creates a data source builder factory using the configuration.
+func NewDataSourceBuilderFactoryFromConfig(dbConfig map[string]interface{}) (physical.DataSourceBuilderFactory, error) {
+	return NewDataSourceBuilderFactory(), nil
+}
+
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables) (execution.RecordStream, error) {
+	file, err := parquet.OpenFile(ds.path)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't open file")
+	}
+	columns := file.Schema.Columns()
+	n := len(columns)
+	for _, col := range columns {
+		if col.MaxR() > 1 {
+			return nil, fmt.Errorf("column '%s' has nested repeated elements", col)
+		}
+	}
+
+	colIters := make([]*ColStrIter, n)
+	for i, col := range columns {
+		colIters[i] = NewColStrIter(file, col)
+	}
+
+	return &RecordStream{
+		file:            file,
+		isDone:          false,
+		columnIterators: colIters,
+		columns:         columns,
+		alias:           ds.alias,
+	}, nil
+}
+
+type RecordStream struct {
+	file            *parquet.File
+	columnIterators []*ColStrIter
+	columns         []parquet.Column
+	isDone          bool
+	alias           string
+}
+
+func (rs *RecordStream) Close() error {
+	err := rs.file.Close()
+	if err != nil {
+		return errors.Wrap(err, "couldn't close underlying file")
+	}
+
+	return nil
+}
+
+func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
+	if rs.isDone {
+		err := rs.file.Close()
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't close underlying file")
+		}
+		return nil, execution.ErrEndOfStream
+	}
+
+	m := make(map[string]interface{})
+
+	rs.isDone = true
```

Comment:
don't operate on `rs.isDone` directly - instead use some temporary variable and afterwards set `if <variable> { rs.isDone = true}`